src/overlays/actors/ovl_En_Fr/z_en_fr.c

```diff
@@ -27,99 +139,966 @@ const ActorInit En_Fr_InitVars = {
+        frog->isButterflyDrawn = frogIndex == frogIndexButterfly ? true : false;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B8D0.s")
+void EnFr_DeactivateButterfly() {
+    s32 frogIndex;
+    EnFr* frog;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B9D8.s")
+    for (frogIndex = 0; frogIndex < ARRAY_COUNT(sEnFrPointers.frogs); frogIndex++) {
+        frog = sEnFrPointers.frogs[frogIndex];
+        frog->isButterflyDrawn = false;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BB34.s")
+// Frogs 2 song is either predetermined (HP not obtained) or random (HP obtained)
+u8 EnFr_GetNextNoteFrogSong(u8 ocarinaNoteIndex) {
+    if ((gSaveContext.eventChkInf[13] & 1) == 0) {
+        return gFrogsSongPtr[ocarinaNoteIndex];
+    } else {
+        return sOcarinaNotes[(s32)Rand_ZeroFloat(60.0f) % 5];
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BC14.s")
+void EnFr_SetupFrogSong(EnFr* this, GlobalContext* globalCtx) {
+    if (this->frogSongTimer) {
+        this->frogSongTimer--;
+    } else { // Initialize timer for frogs 2 HP
+        this->frogSongTimer = 40;
+        this->ocarinaNoteIndex = 0;
+        func_8010BD58(globalCtx, 0x2F);
+        this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
+        EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
+        this->actionFunc = EnFr_ContinueFrogSong;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BCB4.s")
+s32 EnFr_IsFrogSongComplete(EnFr* this, GlobalContext* globalCtx) {
+    u8 index;
+    u8 ocarinaNote;
+    MessageContext* msgCtx = &globalCtx->msgCtx;
+    u8 ocarinaNoteIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BD94.s")
+    if (this->ocarinaNote == (*msgCtx).unk_E410) { // Using (*msgCtx.unk_E410) was the only way to get it to work???
+        this->ocarinaNoteIndex++;
+        ocarinaNoteIndex = this->ocarinaNoteIndex;
+        if (1) {}
+        if (ocarinaNoteIndex >= 14) { // Frog Song is completed
+            this->ocarinaNoteIndex = 13;
+            return true;
+        }
+        // The first four notes have more frames to receive an input
+        index = ocarinaNoteIndex < 4 ? (s32)ocarinaNoteIndex : 4;
+        ocarinaNote = EnFr_GetNextNoteFrogSong(ocarinaNoteIndex);
+        this->ocarinaNote = ocarinaNote;
+        EnFr_CheckOcarinaInputFrogSong(ocarinaNote);
+        this->frogSongTimer = sTimerFrogSong[index];
+    }
+    return false;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BDFC.s")
+void EnFr_OcarinaMistake(EnFr* this, GlobalContext* globalCtx) {
+    func_80106CCC(globalCtx);
+    this->reward = GI_NONE;
+    func_80078884(NA_SE_SY_OCARINA_ERROR);
+    func_800ED858(0); // Comment from z_en_fu.c: seems to be related to setting instrument type (Disable ocarina)
```

Comment:
you can probably omit this comment, we'll have a better idea of what it does with audio decomp